Xamarin.Forms.Platform.Android/Renderers/MediaElementRenderer.cs

```diff
@@ -0,0 +1,321 @@
+				_view.KeepScreenOn = Element.KeepScreenOn;
+
+				Control.AddView(_view);
+
+				_controller = new MediaController(Context);
+				_controller.SetAnchorView(this);
+				_controller.Visibility = Element.ShowsPlaybackControls ? ViewStates.Visible : ViewStates.Gone;
+				_view.SetMediaController(_controller);
+
+				Element.SeekRequested += SeekRequested;
+				Element.StateRequested += StateRequested;
+
+				UpdateSource();
+			}
+		}
+
+		void MetadataRetrieved(object sender, EventArgs e)
+		{
+			Controller.Duration = _view.DurationTimeSpan;
+			Controller.VideoHeight = _view.VideoHeight;
+			Controller.VideoWidth = _view.VideoWidth;
+		}
+
+		void StateRequested(object sender, StateRequested e)
+		{
+			switch (e.State)
+			{
+				case MediaElementState.Playing:
+					_view.Start();
+					Controller.CurrentState = _view.IsPlaying ? MediaElementState.Playing : MediaElementState.Stopped;
+					break;
+
+				case MediaElementState.Paused:
+					if (_view.CanPause())
+					{
+						_view.Pause();
+						Controller.CurrentState = MediaElementState.Paused;
+					}
+					break;
+
+				case MediaElementState.Stopped:
+					_view.Pause();
+					_view.SeekTo(0);
+
+					Controller.CurrentState = _view.IsPlaying ? MediaElementState.Playing : MediaElementState.Stopped;
+					break;
+			}
+
+			UpdateLayoutParameters();
+			Controller.Position = _view.Position;
+		}
+
+		void SeekRequested(object sender, SeekRequested e)
+		{
+			Controller.Position = _view.Position;
+		}
+
+		void UpdateSource()
+		{
+			if (Element.Source != null)
+			{
+				var uriSource = Element.Source as UriMediaSource;
+				if (uriSource != null)
+				{
+					if (uriSource.Uri.Scheme == "ms-appx")
+					{
+						// video resources should be in the raw folder with Build Action set to AndroidResource
+						string uri = "android.resource://" + Context.PackageName + "/raw/" + uriSource.Uri.LocalPath.Substring(1, uriSource.Uri.LocalPath.LastIndexOf('.') - 1).ToLower();
+						_view.SetVideoURI(global::Android.Net.Uri.Parse(uri));
+					}
+					else if (uriSource.Uri.Scheme == "ms-appdata")
+					{
+						string filePath = string.Empty;
+
+						if (uriSource.Uri.LocalPath.StartsWith("/local"))
+						{
+							filePath = System.IO.Path.Combine(System.Environment.GetFolderPath(System.Environment.SpecialFolder.MyDocuments), uriSource.Uri.LocalPath.Substring(7));
+						}
+						else if (uriSource.Uri.LocalPath.StartsWith("/temp"))
+						{
+							filePath = System.IO.Path.Combine(System.IO.Path.GetTempPath(), uriSource.Uri.LocalPath.Substring(6));
+						}
+						else
+						{
+							throw new ArgumentException("Invalid Uri", "Source");
+						}
+
+						_view.SetVideoPath(filePath);
+
+					}
+					else
+					{
+						if (uriSource.Uri.IsFile)
+						{
+							_view.SetVideoPath(uriSource.Uri.AbsolutePath);
+						}
+						else
+						{
+							_view.SetVideoURI(global::Android.Net.Uri.Parse(uriSource.Uri.AbsoluteUri));
+						}
+					}
+				}
+				else
```

Comment:
Could you please use this pattern matching -> `else if (Element.Source is FileMediaSource fileSource)`